File: challenges/programming/esolang/interpreters/py/ctf/interpreter.py
from .player import Player
from .world import World
from .enums import Tile
from .operator import Operator

from typing import List

class Interpreter:
    __players: List[Player]
    __world: World

    def __init__(self, code: bytes):
        self.__players = []

        grid = [[c for c in row.rstrip().replace(b'\t', b' ')] for row in code.split(b'\n')]
        player_pos = []

        for y, row in enumerate(grid):
            for x, v in enumerate(row):
                if v == Tile.PLAYER.value:
                    player_pos.append((x, y))

        for pos in player_pos:
            self.__players.append(Player(pos))
            grid[pos[1]][pos[0]] = Tile.EMPTY.value
        
        self.__world = World(grid)

    @property
    def input(self) -> str:
        return self.__world.input

    @property
    def output(self) -> str:
        return self.__world.output

    def step(self) -> bool:
        for player in self.__players:
            if player.alive: break
        else: return False

        for player in self.__players:
            if player.step():
                Operator.parse(player, self.__world)

        self.__world.step()

        for player in self.__world.new_players:
            self.__players.append(player)

        self.__world.new_players = []

        return True

    def visualize(self):
        grid = self.__world.raw()

        for player in self.__players:
            if player.alive:
                grid[player.position[1]][player.position[0]] = Tile.PLAYER.value

        for row in grid:
            print(bytes(row).decode('utf8'))

        print()

        for player in self.__players:
            if player.alive:
                print(player)

        print("Input:", self.__world.input)
        print("Output:", self.__world.output, '\n')

File: challenges/programming/esolang/interpreters/js/src/interpreter.ts
import { Environment } from "./environment"
import { Player } from "./player";
import { Position, Tile, TileType } from "./utils";
import { World } from "./world";

export class Interpreter {
    private __code: string;
    private __inputFunction: () => string;
    private __environment: Environment;

    constructor(code: string, inputFunction: () => string) {
        this.__code = code;
        this.__inputFunction = inputFunction;
        this.__environment = this.getEnvironment();
    }

    public get input(): string {
        return this.__environment.input;
    }

    public get output(): string {
        return this.__environment.output;
    }

    public raw() {
        return this.__environment.raw();
    }

    public reset() {
        this.__environment = this.getEnvironment();
    }

    public step(): boolean {
        return this.__environment.step();
    }

    public run(): string {
        while(this.step());
        return this.output;
    }

    public visualize() {
        let raw = this.raw();
        let grid = raw.grid;
        let pointers = raw.pointers;

        pointers.forEach(pointer => grid[pointer.position[1]][pointer.position[0]] = '@');

        grid.forEach(row => console.log(row.join('')));
        pointers.forEach(pointer => console.log(pointer.stack));
        console.log("Input:", this.input);
        console.log("Output:", this.output);
        console.log();
    }

    private getGrid(code: string): TileType[][] {
        return code.split('\n').map(row => row.trimEnd().replace('\t', ' ').split(''));
    }

    private getPlayerPositions(grid: TileType[][]) {
        let playerPositions: Position[] = [];

        grid.forEach((row, y) => row.forEach((tile, x) => {
            if(tile == Tile.PLAYER) playerPositions.push(new Position(x, y));
        }))

        return playerPositions;
    }

    private getEnvironment() {
        let grid = this.getGrid(this.__code);
        let playerPositions = this.getPlayerPositions(grid);

        let players: Player[] = playerPositions.map(position => {
            grid[position.y][position.x] = Tile.EMPTY;
            return new Player(position);
        });

        let world = new World(grid);

        return new Environment(players, world, this.__inputFunction);
    }
}
